Last working branch2

Ahmad-Zyoud opened the pull request:
Added progress bar to all steps in the method run_steps. Made 2 new functions to use when creating the progress bar.

Ahmad-Zyoud committed:
Update filehandler.py

ferchaure committed:
fix dup files and rename variable

Ahmad-Zyoud committed:
Update to filehandler.py

- changed check_new_imputs -> check_new_inputs
- input dir can now be as specific or general as the user wants
- added a progress bar when initiating an instance of the object
- fixed a comment

Ahmad-Zyoud committed:
Update to filehandler.py

Fixed a little bug

Ahmad-Zyoud committed:
New function and updates to others.

- added some comments
- added a new parameter to the class, skip_pattern. By default its empty, user can specifiy files to skip if they contain a pattern in their name - its case sensitive.
- changed the description for the single_file_match parameter (although we need to rethink its use, since **/*.isxd will find multiple files unless the input path is specific to the nearst parent folder of the isxd file)
- Turned de_interleave into a step (some code is no longer needed, I made a comment about it that starts with #TASK

- Need to turn create_diff and project movie into a step as well, since these steps make their own .isxd files too and we will want to remove them after moving on to the extract cells steps.

Ahmad-Zyoud committed:
Update filehandler.py

Ahmad-Zyoud committed:
Update filehandler.py

Ahmad-Zyoud committed:
Update filehandler.py

Ahmad-Zyoud committed:
Update filehandler.py

Progress bar was not working correctly, had to create a new variable - amount of files - because grabbing the length from pairlist unzips it.

Ahmad-Zyoud committed:
Update filehandler.py

Turned DFF and project movies into a step.
Still need to get the remove files for them up in running.

Ahmad-Zyoud committed:
Update filehandler.py

Small fix, had to move this around.

Ahmad-Zyoud committed:
Update filehandler.py

Sorry for spamming,

Made some more small changes to get things running

ferchaure commented on cgk_calcium_tools/filehandler.py:
```diff
@@ -57,39 +60,53 @@ class isx_files_handler:
-        files_patterns: Union[str, list] = ".isx",
-        processing_steps: list = ["PP", "TR", "BP", "MC"],
-        single_file_match: bool = True,
+        files_patterns: Union[str, list] = "**/*.isxd",
```
I don't like having  "DI", "DFF" and "PM" as steps. I understand, in could be an easy way to set up all the steps to run from the start. 

On the other hand, this allow us to create a method like ` run_all_steps` (that internally is just a for loop over all the steps)

ferchaure commented on cgk_calcium_tools/filehandler.py:
```diff
@@ -361,28 +341,39 @@ def get_pair_filenames(self, operation: str) -> Tuple[list, list]:
+        local_processing_steps = self.processing_steps
+        
+        # Remove DI from list, if its first
+        if local_processing_steps[0] == "DI":
```
why?

ferchaure commented on cgk_calcium_tools/filehandler.py:
```diff
@@ -361,28 +341,39 @@ def get_pair_filenames(self, operation: str) -> Tuple[list, list]:
 
-        opi = self.processing_steps.index(operation)
+        # Copy the list of processing steps to a local variable
+        local_processing_steps = self.processing_steps
```
self.processing_steps is a list (not an immutable as an int or str), then any change in local_processing_steps will affect self.processing_steps

ferchaure commented on cgk_calcium_tools/filehandler.py:
```diff
@@ -507,21 +512,30 @@ def remove_output_files(self, op, keep_json: bool = True) -> None:
 
         """
-        paths = self.get_filenames(op)
+        if op == 'DI':
+            paths = self.deinterleave_output_files
+        elif op == "DFF":
+            paths = self.get_results_filenames("dff", op=None)
+        elif op == "PM":
+            paths = self.get_results_filenames("maxdff", op=None)
+        else:
+            paths = self.get_filenames(op)
```
why not adding the new steps into get_filenames?

ferchaure commented on cgk_calcium_tools/filehandler.py:
```diff
@@ -550,12 +564,23 @@ def run_step(
                 )
                 crop_rect_files = self.get_results_filenames("crop_rect.csv", op=op)
+            elif op == "DFF":
+                input = self.get_filenames(op=None)
+                output = self.get_results_filenames("dff", op=None)
+                amount_of_files = len(input)
+                pairlist = zip(input, output)
+            elif op == "PM":
+                input = self.get_results_filenames("dff", op=None)
+                output = self.get_results_filenames("maxdff", op=None)
+                amount_of_files = len(input)
+                pairlist = zip(input, output)
+            if op in ["PP","TR", "BP", "MC"]:
```
same here, adding the new steps in get_results_filenames will improve this code

ferchaure commented on cgk_calcium_tools/filehandler.py:
```diff
@@ -1498,6 +1318,10 @@ def preprocess_step(
     """
+    
+    # Initialize progress bar
+    pb = progress_bar(amount_of_files, 'Preprocessing')
```
amount_of_files is just len(pairlist[0]), or do you plan to use  it in another way?

ferchaure commented on cgk_calcium_tools/filehandler.py:
```diff
@@ -1717,3 +1866,30 @@ def parameters_for_isx(
     return copy_dict
+
+
+def progress_bar(end_amount, description_step):
+    # Initialize progress bar
+    progress_bar = IntProgress(
+        value=0, 
+        min=0,
+        max=end_amount, 
+        style={
+            'bar_color': '#3385ff',
+            'description_width': 'auto'
+            },
+        description= f'{description_step} Movies: 0/{end_amount} movie(s)',
+        layout=Layout(width='45%')
+        )
+            
+    # Display the progress bar
+    display(progress_bar)
+    return progress_bar
+
+
+def update_progress_bar(progress_bar, description_step):
+    progress_bar.value += 1
+    if progress_bar.value == progress_bar.max:
+        progress_bar.description = f'{description_step} Movies: {progress_bar.value}/{progress_bar.max} Complete!'
+    else:
+        progress_bar.description = f'{description_step} Movies: {progress_bar.value}/{progress_bar.max} movie(s)'
```
I would re implement the first function as an constructor class and the second as its method. In that way you don't have to pass two times `description_step` .

can you move it to another file? maybe `jupyter_outputs.py ` ?

ferchaure reviewed (changes requested):
Amazing work!!!

With a few extra changes, we can make the code quite a bit simpler.


Thanks for all the comments

Ahmad-Zyoud committed:
Update filehandler.py

Added a way to measure the time it takes to run a method. Could help us estimate how long it takes to process a number of files.

Ahmad-Zyoud committed:
Progress bar updates

Progress bar is now a class
new file - jupyter_outputs.py

Ahmad-Zyoud committed:
Debugging Filehandler

- Better return value for De_interleave
- Fixed skip_pattern so it works when there is no input.

Ahmad-Zyoud committed:
Update jupyter_outputs.py

updated the progress bar width to be fixed regardless of description length.

ferchaure commented on cgk_calcium_tools/filehandler.py:
```diff
@@ -24,29 +27,42 @@ def ifstr2list(x) -> list:
     return [x]
 
 
+def timer(method):
+    def timed(self, *args, **kwargs):
+        start_time = perf_counter()
+        result = method(self, *args, **kwargs)
+        end_time = perf_counter()
+        elapsed_time = end_time - start_time
+        print(f'{method.__name__} executed in {timedelta(seconds=elapsed_time)}')
+        self.total_time += elapsed_time
+        return result
+    return timed
```
sorry, I don't get it , usually the notebooks tell you how much time is used in each cell.

Ahmad-Zyoud committed:
Changes to files_io

Some minor changes to progress bar.

I had to move the create_inscopix_projects function to filehandler.py because there is circular import loop between filehandler and files_io. Filehandler imports files_io while files_io imports filehandler

Ahmad-Zyoud committed:
Update filehandler.py

Changes to get the file hack to work.
- path file names need to have a forward slash.
- data_folder must not include a "." so changed cropped string to -5 instead of -4

Ahmad-Zyoud committed:
progress bars

more progress bars

Ahmad-Zyoud committed:
Fixed multiplane reg

Multiplane reg works now

Don't worry changed the variable name from temp to input_cell_set_file_names

Ahmad-Zyoud committed:
Progress bar for multiplane registration

- pb for multiplane registration
- also started a deconvolve function, not functional yet

Ahmad-Zyoud committed:
Increased block size + added comment
